src/client.rs

```diff
@@ -68,29 +82,62 @@ impl<S: Service, C: ServiceConnection<S>, T: Into<S::Req>> Sink<T> for UpdateSin
     }
 }
 
-impl<S: Service, C: ServiceConnection<S>> RpcClient<S, C> {
+impl<S, C> RpcClient<S, C, S>
+where
+    S: Service,
+    C: ServiceConnection<S>,
+{
     /// Create a new rpc client for a specific [Service] given a compatible
     /// [ServiceConnection].
     ///
     /// This is where a generic typed connection is converted into a client for a specific service.
     pub fn new(source: C) -> Self {
         Self {
             source,
-            p: PhantomData,
+            map: Arc::new(Mapper::new()),
         }
     }
+}
 
+impl<S, C, S2> RpcClient<S, C, S2>
+where
+    S: Service,
+    C: ServiceConnection<S>,
+    S2: Service,
+{
     /// Get the underlying connection
     pub fn into_inner(self) -> C {
         self.source
     }
 
+    /// Map this channel's service into an inner service.
+    ///
+    /// This method is available if the required bounds are upheld:
+    /// S3::Req: Into<S2::Req> + TryFrom<S2::Req>,
+    /// S3::Res: Into<S2::Res> + TryFrom<S2::Res>,
+    ///
+    /// Where S3 is the new service to map to and S2 is the current inner service.
+    ///
+    /// This method can be chained infintely.
+    pub fn map<S3>(self) -> RpcClient<S, C, S3>
+    where
+        S3: Service,
+        S3::Req: Into<S2::Req> + TryFrom<S2::Req>,
+        S3::Res: Into<S2::Res> + TryFrom<S2::Res>,
+    {
+        let map = ChainedMapper::new(self.map);
+        RpcClient {
+            source: self.source,
+            map: Arc::new(map),
```

Comment:
This needs to be Arc so it can be cloned? Box won't do?

Reply:
Yes, at least in the client it needs to be cloneable because the client needs to be cloneable.